azurelinuxagent/common/protocol/wire.py

```diff
@@ -1195,14 +1196,73 @@ def get_artifacts_profile(self):
                     msg = "Content: [{0}]".format(profile)
                     logger.verbose(msg)
 
-                    from azurelinuxagent.common.event import report_event, WALAEventOperation
                     report_event(op=WALAEventOperation.ArtifactsProfileBlob,
                                  is_success=False,
                                  message=msg,
                                  log_event=False)
 
         return artifacts_profile
 
+    def upload_logs(self, content):
+        try:
+            host_plugin = self.get_host_plugin()
+            host_plugin.put_vm_log(content)
+
+            msg = "Upload VM logs request succeeded using the host plugin channel for " \
+                  "container id {0} and role config file {1}".format(host_plugin.container_id,
+                                                                     host_plugin.role_config_name)
+            add_event(name=AGENT_NAME,
+                      version=CURRENT_VERSION,
+                      op=WALAEventOperation.HostPlugin,
+                      is_success=True,
+                      message=msg,
+                      log_event=True)
+        except (ResourceGoneError, InvalidContainerError) as e:
+            host_plugin = self.get_host_plugin()
+            old_container_id = host_plugin.container_id
+            old_role_config_name = host_plugin.role_config_name
+
+            msg = "Upload VM logs request failed with the current host plugin configuration. " \
+                  "ContainerId: {0}, role config file: {1}. Fetching new goal state and retrying the call." \
+                  "Error: {2}".format(old_container_id, old_role_config_name, ustr(e))
+            logger.info(msg)
+
+            self.update_host_plugin_from_goal_state()
+
+            new_container_id = host_plugin.container_id
+            new_role_config_name = host_plugin.role_config_name
+            msg = "Host plugin reconfigured with new parameters. " \
+                  "ContainerId: {0}, role config file: {1}.".format(new_container_id, new_role_config_name)
+            logger.info(msg)
+
+            try:
+                host = self.get_host_plugin()
```

Comment:
there is too much code repetition here

Reply:
I thought about refactoring this into its own little method:
```
def _get_host_and_put_logs(content, msg):
    host = self.get_host_plugin()
    host.put_vm_log(content)

    logger.info(msg)
    add_event(name=AGENT_NAME,
                version=CURRENT_VERSION,
                op=WALAEventOperation.HostPlugin,
                is_success=True,
                message=msg,
                log_event=False)
```
I didn't because I didn't think it added a lot of value. If you this it does or you have better suggestions how to make this method more readable, please let me know!

Reply:
Suggestion: Can't you make the function a lambda and make the whole retry-and-update-host-logic a helper function that you use here and in the send_appropriate_request_function too?

Reply:
I think Joe added a helper function like that in his PR. You could take a look at that too for inspiration

Reply:
we have multiple versions of the "try the host plugin, catch resource gone/invalid container, get a new goal state, update the plugin, and now retry" logic. it'd be good to capture that logic in a single method in the plugin and use it from the other places. 

or use joe's function if appropriate

Reply:
Used Joe's approach.